src/masternodes/govvariables/attributes.cpp

```diff
@@ -312,6 +362,65 @@ Res ATTRIBUTES::ProcessVariable(const std::string& key, const std::string& value
     return Res::Err("No parse function {%d, %d}", type, typeKey);
 }
 
+Res ATTRIBUTES::RefundFuturesContracts(CCustomCSView &mnview, const uint32_t height, const uint32_t tokenID)
+{
+    CDataStructureV0 blockKey{AttributeTypes::Param, ParamIDs::DFIP2203, DFIPKeys::BlockPeriod};
+    const auto blockPeriod = GetValue(blockKey, CAmount{});
+    if (blockPeriod == 0) {
+        return Res::Ok();
+    }
+
+    const uint32_t startHeight = height - (height % blockPeriod);
+    std::map<CFuturesUserKey, CFuturesUserValue> userFuturesValues;
+
+    mnview.ForEachFuturesUserValues([&](const CFuturesUserKey& key, const CFuturesUserValue& futuresValues) {
+        if (key.height <= startHeight) {
+            return false;
+        }
+
+        if (tokenID != std::numeric_limits<uint32_t>::max()) {
+            if (futuresValues.source.nTokenId.v == tokenID || futuresValues.destination == tokenID) {
+                userFuturesValues[key] = futuresValues;
+            }
+        } else {
+            userFuturesValues[key] = futuresValues;
+        }
+
+        return true;
+    }, {height, {}, std::numeric_limits<uint32_t>::max()});
+
+    const auto resVal = GetFutureSwapContractAddress();
+    if (!resVal) {
+        return resVal;
+    }
+
+    CDataStructureV0 liveKey{AttributeTypes::Live, ParamIDs::Economy, EconomyKeys::DFIP2203Tokens};
+    auto balances = GetValue(liveKey, CBalances{});
+
+    for (const auto& [key, value] : userFuturesValues) {
+        mnview.EraseFuturesUserValues(key);
+
+        auto res = mnview.SubBalance(*resVal, value.source);
+        if (!res) {
+            return res;
+        }
+
+        res = mnview.AddBalance(key.owner, value.source);
+        if (!res) {
+            return res;
+        }
```

Comment:
why not use transferTokenBalance?

Reply:
It is not available there, currently a member of CCustomTxVisitor which uses members of its class. Would have to be moved out and all current uses of it updated.